### ISSUE TITLE
Add Password Input to User Resource

### ISSUE THREAD

**Issue by codingdiaz** [atlassian/terraform-provider-artifactory#27]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

When creating a user resource it would be useful to be able to specify a password. This could be a randomly generated password. How exactly would you get the password after creating a user? (assuming that an admin would have to login and reset the password the way this currently works?)

### New or Affected Resource(s)

NA

### Potential Terraform Configuration

```
# Create a new Artifactory user called terraform
resource "artifactory_user" "test-user" {
  name   = "terraform"
  email  = "test-user@artifactory-terraform.com"
  groups = ["logged-in-users", "readers"]
  password = "somerandompassword"
}
```

### References

<!---
NA

* #0000

**Comment by dillon-giacoppo** [comment]
This is already possible by setting the "TF_USER_%s_PASSWORD" in the environment where %s is the user name field. Terraform cannot manage user passwords since Artifactory returns their encrypted version. Therefore this value only checks/works on create, it does not update the value if the password changes. This is a limitation of Artifactory API and not the provider unfortunately. I will update the docs to make this environment variable more clear

**Comment by codingdiaz** [comment]
That makes sense. Where exactly are the docs? I found this but, couldn't find much else. https://github.com/atlassian/terraform-provider-artifactory/blob/master/website/docs/r/artifactory_user.html.markdown

Would there be any way to retrieve this value? https://github.com/atlassian/terraform-provider-artifactory/blob/master/pkg/artifactory/resource_artifactory_user.go#L109

**Comment by Constantin07** [comment]
@dillon-giacoppo Are you sure that the password is encrypted and not just hashed somehow using publically avalable algorithms ? If the last, JFrog could probably share this information, I guess.

**Comment by dillon-giacoppo** [comment]
@codingdiaz User passwords are never returned through the API. Since they are never returned they cannot be managed by terraform. Replication and remote repo passwords do get returned so they can be fully managed if encryption is disabled.

@Constantin07 See https://www.jfrog.com/confluence/display/RTF/Artifactory+Key+Encryption, they are definitely encrypted not hashed. In the next week, I will be investigating a possible solution where we store the encrypted hash and the plaintext hash together like `plaintext::encrypted` and do a diff on either value depending on the operation. I am not sure how feasible this is so can't promise it will make it into v2 (which should be released next week). In an ideal world the encryption/decryption would be publicly available and we could do offline decryption with the masterkey.  AFAIK there is no public documentation on this process.

**Comment by codingdiaz** [comment]
Yep, I looked yesterday and you can't get that back.

**Comment by Code0x58** [comment]
One fun thing with the `TF_USER_%s_PASSWORD` approach to setting the password is that it only works with a subset of valid usernames. Switching to use resource names also wouldn't fix it was [identifiers](https://www.terraform.io/docs/configuration/syntax.html#identifiers) can have hyphens (although maybe a better limitation).

I feel like it would be possible to have a password input which is only used during creation, then you could update your tfvars and taint the affected users, then run apply (twice, because of auto-joined groups).

This approach if doable would be forward compatible with any Read/Update related password changes down the line.

_**disclaimer:** I haven't read the terraform custom provider documentation in a while_

**Comment by dillon-giacoppo** [comment]
I see, a quick fix is to base64 encode the username (and probably should do the password as well). I will look at adding a `TF_USER_%s_PASSWORD_ENC` environment variable

**Comment by dillon-giacoppo** [comment]
So I am not sure Artifactory supports special characters in usernames, but regardless, v1.5 will support variables that have an MD5 username and a Base64 password identified by an additional `_ENC` suffix. Base64 cannot be used for the environment var name since it pads with illegal '=' character. By doing this you can use whatever UTF-8 characters you like. Original usage will continue to be supported as well.

Example usage:
```bash
# Old style password environment variables
export "TF_USER_testuser_PASSWORD"="testpassword"

# Hash the username and store the password as base64.
export "TF_USER_$(echo -n "testuser" | md5)_PASSWORD_ENC"="$(echo -n "testpassword" | base64)"
```

**Comment by Code0x58** [comment]
Thanks, the kind of username I encountered the issue with was `pypi-consumer` (I don't know if there's unicode support).

Is there a reason to add base64 encoding to the password? It introduces a minor possible risk of erroneously non-encoded passwords being valid base64 so making a nerfed user, when I'm not sure it provides a feature.

In the evening I might check if having the password in the tf file is doable and make a PR if so as it'd be nice to avoid wrapper scripts/environs.

**Comment by dillon-giacoppo** [comment]
Passwords have a wide range of possible special characters (consider how the original environment variable would handle spaces). It is painful to escape them and unescape them by hand. Base64 is a simple way of getting them in and out of the environment which is quite universal. I am skeptical of any "in-resource" solutions since that is not how attributes are intended to be used and creates behaviour that is inconsistent with other terraform providers. I don't believe your use-case justifies more than environment variable.

For any proposal to be accepted it must be an improvement on existing functionality. Any design for an in-TF-file password will need to be discussed and justified.

I would look to see how other providers handle such a situation before building a from scratch solution. Lastly, if you do decide to try a proof of concept and experiment, please build against the v1.5 branch and ensure you do not break existing configurations.

**Comment by Code0x58** [comment]
Values in the environment generally aren't limited (aside from no null bytes), it's usually just the keys that have limitations, which is worked around with the md5sum.

I imagine the tf based password approach be a minor change using [schema behaviours](https://www.terraform.io/docs/extend/schemas/schema-behaviors.html) which would add one for `artifactory_user.password`. Then users would do something like `password = "${var.my_user_password}"` or `password = "${data.some_provider.my_instance.my_user_password}"`. I think there are other places where the DiffFunc always returns false as there's no API to retrieve the state, e.g. password.

**Comment by jamestoyer** [comment]
Closed by #42
